astropy/cosmology/flrw/base.py

```diff
@@ -356,66 +312,77 @@ def m_nu(self, param, value):
     # properties
 
     @property
-    def is_flat(self):
+    def is_flat(self) -> bool:
         """Return bool; `True` if the cosmology is flat."""
-        return bool((self._Ok0 == 0.0) and (self.Otot0 == 1.0))
+        return bool((self.Ok0 == 0.0) and (self.Otot0 == 1.0))
 
     @property
-    def Otot0(self):
+    def Otot0(self) -> float:
         """Omega total; the total density/critical density at z=0."""
-        return self._Om0 + self._Ogamma0 + self._Onu0 + self._Ode0 + self._Ok0
+        return self._Om0 + self.Ogamma0 + self.Onu0 + self._Ode0 + self.Ok0
 
-    @property
-    def Odm0(self):
+    @cached_property
+    def Odm0(self) -> float | None:
         """Omega dark matter; dark matter density/critical density at z=0."""
-        return self._Odm0
+        return None if self.Ob0 is None else (self.Om0 - self.Ob0)
 
-    @property
-    def Ok0(self):
+    @cached_property
+    def Ok0(self) -> float:
         """Omega curvature; the effective curvature density/critical density at z=0."""
-        return self._Ok0
+        return 1.0 - self.Om0 - self.Ode0 - self.Ogamma0 - self.Onu0
 
-    @property
-    def Tnu0(self):
+    @cached_property
+    def Tnu0(self) -> u.Quantity:
         """Temperature of the neutrino background as |Quantity| at z=0."""
-        return self._Tnu0
+        # The constant in front is (4/11)^1/3 -- see any cosmology book for an
+        # explanation -- for example, Weinberg 'Cosmology' p 154 eq (3.1.21).
+        return 0.7137658555036082 * self.Tcmb0
 
     @property
-    def has_massive_nu(self):
+    def has_massive_nu(self) -> bool:
         """Does this cosmology have at least one massive neutrino species?"""
-        if self._Tnu0.value == 0:
+        if self.Tnu0.value == 0:
             return False
         return self._massivenu
 
-    @property
-    def h(self):
+    @cached_property
+    def h(self) -> float:
         """Dimensionless Hubble constant: h = H_0 / 100 [km/sec/Mpc]."""
-        return self._h
+        return self.H0.value / 100.0
 
-    @property
-    def hubble_time(self):
+    @cached_property
+    def hubble_time(self) -> u.Quantity:
```

Comment:
Let's try.
I think we might need to fix some stuff in the type annotations to `Quantity` for this to work as static annotations.

Reply:
![CleanShot 2024-07-11 at 14 14 36@2x](https://github.com/astropy/astropy/assets/8949649/7c03a7fd-0826-47c8-b509-feb8ad1039b8)

We'd have to reconfigure `ruff` to work with string annotations.

Reply:
Type checkers might not be able to understand physical types, but we are not type checking yet and surely this is useful for human readers.

Reply:
IMO it's better to get `u.Quantity` annotations in now and have a collective think about the future of annotations for Quantity. e.g. will dimensions be numpy dtypes? Or ...

Reply:
It would be very useful to be able to specify the physical type of a `Quantity` in type annotations, but I can see now that it is not as simple to write it down as I hoped it would be, so it's beyond the scope for this pull request.